### PR TITLE
lint/param-name-checker: duplication checker v1

### PR DESCRIPTION
**Pull request by fexolm**

### CODE REVIEW

**Comment by Quasilyte** [comment]
I'll take a look a bit later.

**Comment by fexolm** [comment]
@Quasilyte i have moved it to sepatate file.

File: lint/param-duplication_checker.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+// could be simplified by combine arguments with same type
+//
+// Example: func f(a int, b int) could be simplified as func f(a, b int)
+type ParamDuplicationChecker struct {
+	ctx *Context
+
+	warnings []Warning
+}
+
+// NewParamDuplicationChecker returns initialized ParamDuplicationChecker
```
Comment (Quasilyte) [inline lint/param-duplication_checker.go]:
```
// NewParamDuplicationChecker returns initialized checker for function param types.
```
Note the trailing period.

File: lint/param-duplication_checker.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+import (
+	"fmt"
+	"go/ast"
+
+	"github.com/Quasilyte/astcmp"
+)
+
+// ParamDuplicationChecker detects functions where parameters declaration
+// could be simplified by combine arguments with same type
```
Comment (Quasilyte) [inline lint/param-duplication_checker.go]: Missing period.

File: lint/param-duplication_checker.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+// TODO(fexolm) don't create multiple warnings on the same function
+// TODO(fexolm) create warning in other function
+func (c *ParamDuplicationChecker) checkParamDuplication(decl *ast.FuncDecl) {
+	params := decl.Type.Params.List
+	if len(params) < 2 {
+		return
+	}
+	for i, p := range params[1:] {
+		if astcmp.EqualExpr(p.Type, params[i].Type) {
+			var winfo string
```
Comment (Quasilyte) [inline lint/param-duplication_checker.go]: We'll find some way to make message construction easier later.
OK for now.
Thought you could move whole block into it's own method, like `warn`.

File: lint/util.go [hunk 4]
```diff
@@ -24,3 +25,14 @@ func nodeString(fset *token.FileSet, x ast.Node) string {
 	}
 	return buf.String()
 }
+
+func paramNamesStr(idents []*ast.Ident) string {
```
Comment (Quasilyte) [inline lint/util.go]: Can you make this a method of `ParamDuplicationChecker`?

File: lint/param-duplication_checker.go (hunk 3 above)
Comment (fexolm) [reply lint/param-duplication_checker.go]: I have TODO for this :)

File: lint/param-duplication_checker.go [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+}
+
+// NewParamDuplicationChecker returns initialized ParamDuplicationChecker
+func NewParamDuplicationChecker(ctx *Context) *ParamDuplicationChecker {
+	return &ParamDuplicationChecker{
+		ctx: ctx,
+	}
+}
+
+// Check runs parameter duplication checks for f.
```
Comment (Quasilyte) [inline lint/param-duplication_checker.go]:
```diff
- Check runs parameter duplication checks for f.
+ Check runs parameter types duplication checks for f.
```

File: lint/util.go (hunk 4 above)
Comment (fexolm) [reply lint/util.go]: I thought it could be useful for other chckers. It looks quite general :)

File: lint/param-duplication_checker.go (hunk 3 above)
Comment (Quasilyte) [reply lint/param-duplication_checker.go]: Do you have particular reasons for this?

I do believe that TODO are only good for things that you are not going to do in a few days or hours. If it's a matter of moving code into a method, it's better to do it right away, at least that's my preference.

File: lint/util.go (hunk 4 above)
Comment (Quasilyte) [reply lint/util.go]: We'll move it to `util.go` as soon as there is a second function user. :)
Let's be conservative about what makes it's way to `util.go`.

Comment (fexolm) [reply lint/util.go]: Ok

File: lint/param-duplication_checker.go (hunk 3 above)
Comment (fexolm) [reply lint/param-duplication_checker.go]: Done

File: lint/util.go (hunk 4 above)
Comment (fexolm) [reply lint/util.go]: Done

**Comment by Quasilyte** [comment]
Dat `interfacer`. :)

**Comment by fexolm** [comment]
CI failed with `decl can be go/ast.Node (interfacer)`. I think we could merge that.

**Comment by Quasilyte** [comment]
@fexolm, yep.

**Pull request merged**